Add _size and _doc_count to fields output

cbuescher opened the pull request:
Currently metadata fields like `_size` or `_doc_count` cannot be retrieved using
the fields API. With this change, we allow this if the field is explicitely
queried for using its name, but won't include metadata fields when e.g.
requesting all fields via "*".
With this change, not all metadata fields will be retrievable by using its name,
but support for "_size" and "_doc_count" (which is fetched from source) is
added. Support for other metadata field types will need to be decided case by
case and an appropriate ValueFetcher needs to be supplied.

Relates to #63569

Commit:
Add _size and _doc_count to fields output

Currently metadata fields like `_size` or `_doc_count` cannot be retrieved using
the fields API. With this change, we allow this if the field is explicitely
queried for using its name, but won't include metadata fields when e.g.
requesting all fields via "*".
With this change, not all metadata fields will be retrievable by using its name,
but support for "_size" and "_doc_count" (which is fetched from source) is
added. Support for other metadata field types will need to be decided case by
case and an appropriate ValueFetcher needs to be supplied.

Relates to #63569

elasticmachine commented:
Pinging @elastic/es-search (Team:Search)

Commit:
fix test

Commit:
Merge branch 'master' into fields-metadata

jtibshirani reviewed (commented):
This looks good to me! It'd be nice to update `mapper-size.asciidoc` with this simplified approach, maybe removing references to `script_fields` and `docvalue_fields`.

Commit:
Merge branch 'master' into fields-metadata

Commit:
iter

cbuescher commented:
@jtibshirani thanks, I added the suggested small changes. I left the script example in the docs since I find it useful to also mention how the field can be accessed in scripts, but I replaced the doc values retrieval example with one using the fields API.

Commit:
iter

cbuescher commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into fields-metadata

jtibshirani commented on docs/plugins/mapper-size.asciidoc:
```diff
@@ -85,9 +81,4 @@ GET my-index-000001/_search
 <4> Uses a
```
Do we want to remove this reference to `script_fields` too? It doesn't seem so useful, and I think we'll be phasing out `script_fields` now that we have runtime fields.

cbuescher replied:
I left it here to show an example of general access in scripts, the `script_field` seems to provide a simple way of giving that example. But I can remove it if you don't find it useful. I think I do.

jtibshirani replied:
No strong preference from me. As a compromise we could just swap the order to emphasize `fields` is the preferred way to retrieve its value?

cbuescher replied:
Sounds like a good idea.

jtibshirani reviewed (commented):
I noticed we merged the new `_tier` metadata field. Do we plan to handle that in this PR, or a follow-up?

cbuescher commented:
> I noticed we merged the new _tier metadata field. Do we plan to handle that in this PR, or a follow-up?

I saw the issue, wasn't sure it was merged already. And yes, if we want to support that meta data field I'd prefer this being a follow up, seeing this PR is close to merging I prefer not to include it.

jtibshirani commented on docs/plugins/mapper-size.asciidoc:
```diff
@@ -26,7 +26,7 @@ PUT my-index-000001
 
 The value of the `_size` field is accessible in queries, aggregations, scripts,
-and when sorting:
+and when sorting. It can be retrieved using the {ref}/search-fields.html#search-fields-param[fields API]:
```
I just noticed this: we've been trying not to call it the 'fields API' in official docs. It has caused confusion in the past because some users thought it was a separate API endpoint. We could say `fields` parameter or `fields` option instead.

jtibshirani replied:
Oops, not sure this was updated before merging?

cbuescher replied:
Indeed, I missed that one, only changed the mention below in the example. Will push a small online-fix.

cbuescher replied:
Corrected this.

jtibshirani reviewed (approved):
I left a couple last documentation suggestions, once those are addressed it looks ready to merge!

Commit:
docs changes

cbuescher commented:
@jtibshirani given the GA discussion, do you think this should also go into 7.12.x?

jtibshirani commented:
I don't think it needs to go into 7.12, since it is an enhancement and doesn't change the API. We made sure to list it under 'other changes' on the meta issue (#60985) to clarify it doesn't affect the API.
